### PR TITLE
Create an index file while writing a sorted BAM file

### PR DESCRIPTION
**Pull request by xckitahara**
This PR is to add a function to create an index file if it is already sorted when creating a BAM file.
PR for issue #141 .

### CODE REVIEW

**Comment by alumi** [comment]
Hi @xckitahara, Thank you for working on this feature!
Could you squash your commits into some concise and meaningful ones and then rebase them onto the master branch?

**Comment by alumi** [comment]
It seems like GitHub cannot associate your commits with your account @xckitahara.
Please configure your git client with [`git config --global user.email`](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#_your_identity).

**Comment by xckitahara** [comment]
> Hi @xckitahara, Thank you for working on this feature!
> Could you squash your commits into some concise and meaningful ones and then rebase them onto the master branch?

@alumi Has been updated.

**Comment by xckitahara** [comment]
> It seems like GitHub cannot associate your commits with your account @xckitahara.
> Please configure your git client with [`git config --global user.email`](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup#_your_identity).

I'm sorry.
Set up.

**Comment by alumi** [comment]
> Set up
Great :+1:
Now you need to amend `—reset-author` commits to fix the committer and the author

File: src/cljam/io/sam.clj [hunk 1]
```diff
@@ -89,17 +89,26 @@
 (defn ^BAMWriter bam-writer
   "Returns an open cljam.io.bam.writer.BAMWriter of f. Should be used inside
   with-open to ensure the writer is properly closed."
-  [f]
-  (bam-core/writer f))
+  ([f]
+   (bam-core/writer f false))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/sam.clj]: I fix it.

File: src/cljam/io/bam/writer.clj [hunk 2]
```diff
@@ -1,28 +1,54 @@
...
-    (.close ^Closeable (.writer this)))
+    (let [last-pointer (.getFilePointer ^BGZFOutputStream (.writer this))]
+      (.close ^Closeable (.data-writer this))
+      (when @(.index? this)
+        (let [last-pointer' (.getFilePointer ^BGZFOutputStream (.writer this))
+              index' (bai-writer/update-lastpointer-index (count @(.refs this)) @(.index this) last-pointer last-pointer')]
+          (with-open [w (DataOutputStream. (FileOutputStream. (cio/file (str (.filename this) ".bai"))))]
+            (let [nrefs (count @(.refs this))
+                  indices (bai-writer/finalize-index nrefs index')]
+              (bai-writer/write-index*! w nrefs indices)))))))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: I fix it.

File: src/cljam/io/sam.clj [hunk 3]
```diff
@@ -89,17 +89,26 @@
...
-  [f]
-  (case (io-util/file-type f)
-    :sam (sam-writer f)
-    :bam (bam-writer f)
-    (throw (IllegalArgumentException. "Invalid file type"))))
+  ([f]
+   (case (io-util/file-type f)
+     :sam (sam-writer f)
+     :bam (bam-writer f)
+     (throw (IllegalArgumentException. "Invalid file type"))))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/sam.clj]: I fix it.

File: src/cljam/io/bam/core.clj [hunk 4]
```diff
@@ -52,6 +52,12 @@
...
-              (DataOutputStream. (bgzf/bgzf-output-stream f))))
+(defn ^BAMWriter writer [f create-index?]
+  (let [w (bgzf/bgzf-output-stream f)]
+    (BAMWriter. (util/as-url f)
+                w
+                (DataOutputStream. w)
+                f
+                (atom create-index?)
+                (atom nil)
+                (atom nil))))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/core.clj]: I fix it.

File: src/cljam/io/bam/writer.clj [hunk 5]
```diff
@@ -1,28 +1,54 @@
...
   (writer-url [this]
     (.url this))
   protocols/IAlignmentWriter
   (write-header [this header]
+    (when @(.index? this)
+      (if (sort-by-pos? header)
+        (do
+          (reset! (.refs this) (refs/make-refs header))
+          (reset! (.index this) {:no-coordinate-alns 0}))
+        (reset! (.index? this) false)))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: I fix it.

File: src/cljam/io/bam/writer.clj [hunk 6]
```diff
@@ -1,28 +1,54 @@
...
+              (bai-writer/write-index*! w nrefs indices)))))))
   protocols/IWriter
   (writer-url [this]
     (.url this))
   protocols/IAlignmentWriter
   (write-header [this header]
+    (when @(.index? this)
+      (if (sort-by-pos? header)
+        (do
+          (reset! (.refs this) (refs/make-refs header))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: I fix it.

File: src/cljam/io/bam_index/writer.clj [hunk 7]
```diff
@@ -310,7 +228,127 @@
...
+                             (zipmap bin-keys (map update-fn ch-vals))))]
+    (loop [i 0
+           index index]
+      (if (< i nrefs)
+        (if (get index i)
+          (recur (inc i) (-> index
+                             (update-in [i :meta-data] (partial update-pt-fn-meta oldpt newpt))
+                             (update-in [i :bin-index] (partial update-pt-fn-bin oldpt newpt))))
+          (recur (inc i) index))
+        index))))
```
Comment (alumi) [inline src/cljam/io/bam_index/writer.clj:314]: https://github.com/bbatsov/clojure-style-guide#higher-order-functions

Don't use `loop`-`recur` unless you need to optimize the function really heavily.
We can use higher order functions and macros instead to improve readability like the followings:

```clojure
(defn update-last-pointer [index last-buf-ptr eof-ptr]
  (reduce
   (fn [r [ref-id bin i]]
     (-> r
         (assoc-in [ref-id :bin-index bin i :end] eof-ptr)
         (assoc-in [ref-id :meta-data :last-offset] eof-ptr)))
   index
   (for [[ref-id {:keys [bin-index]}] (dissoc index :no-coordinate-alns)
         [bin chunks] bin-index
         [i {:keys [end]}] (map-indexed vector chunks)
         :when (= end last-buf-ptr)]
     [ref-id bin i])))
```

or maybe we can write it in a more direct way:

```clojure
(defn update-last-pointer [index eof-ptr]
  (if (pos? (get index :no-coordinate-alns 0))
    index
    (let [last-ref (apply max (keys (dissoc index :no-coordinate-alns)))
          last-key (->> (for [[bin chunks] (get-in index [last-ref :bin-index])
                              [i {:keys [end]}] (map-indexed vector chunks)]
                          [end [last-ref :bin-index bin i :end]])
                        (apply max-key first)
                        last)]
      (-> index
          (assoc-in [last-ref :meta-data :last-offset] eof-ptr)
          (assoc-in last-key eof-ptr)))))
```

(Please correct me if I'm wrong 🙏 )

File: src/cljam/io/bam/writer.clj [hunk 8]
```diff
@@ -1,23 +1,45 @@
...
+            [cljam.io.bam.decoder :as bam-decoder]
+            [cljam.io.bam-index.writer :as bai-writer])
+  (:import [java.io Closeable DataOutputStream]
+           [cljam.io.bam.decoder BAMRawBlock]
+           [bgzf4j BGZFOutputStream]))
 
 (declare write-header* write-refs* write-alignments* write-blocks*)
 
+(defn- sort-by-pos?
+  "Returns true if the bam is sorted by coordinate, false if not. It is detected by
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj [hunk 9]
```diff
@@ -53,14 +78,38 @@
...
+                               (.reset baos)
+                               (encoder/encode-alignment dos a refs)
+                               (lsb/write-int dw (.size baos))
+                               (.writeTo baos dw)
+                               (when @(.index wtr)
+                                 (bam-decoder/decode-pointer-block
+                                  (BAMRawBlock. (.toByteArray baos) pointer-beg (.getFilePointer w))))))
+                           alns)]
+        (if @(.index wtr)
+          (reset! (.index wtr) (bai-writer/merge-index @(.index wtr) (bai-writer/make-index* pointer-block)))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj [hunk 10]
```diff
@@ -53,14 +78,38 @@
...
+                       (fn [{:keys [data]}]
+                         (let [pointer-beg (.getFilePointer w)]
+                           (lsb/write-int dw (alength ^bytes data))
+                           (lsb/write-bytes dw data)
+                           (when @(.index wtr)
+                             (bam-decoder/decode-pointer-block
+                              (BAMRawBlock. data pointer-beg (.getFilePointer w))))))
+                       blocks)]
+    (if @(.index wtr)
+      (reset! (.index wtr) (bai-writer/merge-index @(.index wtr) (bai-writer/make-index* pointer-block)))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj [hunk 11]
```diff
@@ -53,14 +78,38 @@
...
+      (doall pointer-block))))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj [hunk 12]
```diff
@@ -53,14 +78,38 @@
...
+                               (encoder/encode-alignment dos a refs)
+                               (lsb/write-int dw (.size baos))
+                               (.writeTo baos dw)
+                               (when @(.index wtr)
+                                 (bam-decoder/decode-pointer-block
+                                  (BAMRawBlock. (.toByteArray baos) pointer-beg (.getFilePointer w))))))
+                           alns)]
+        (if @(.index wtr)
+          (reset! (.index wtr) (bai-writer/merge-index @(.index wtr) (bai-writer/make-index* pointer-block)))
+          (doall pointer-block))))))
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam_index/writer.clj (hunk 7 above)
Comment (xckitahara) [reply src/cljam/io/bam_index/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj [hunk 13]
```diff
@@ -1,23 +1,45 @@
...
-    (.close ^Closeable (.writer this)))
+    (let [last-pointer (.getFilePointer ^BGZFOutputStream writer)]
+      (.close ^Closeable (.data-writer this))
+      (when-let [idx @index]
+        (let [last-pointer' (.getFilePointer ^BGZFOutputStream writer)
+              n-refs (count @refs)]
+          (with-open [w (->> (str url ".bai")
+                             cio/output-stream
+                             DataOutputStream.)]
+            (->> (bai-writer/update-lastpointer-index n-refs idx last-pointer last-pointer')
```
[the remark this answers is not in the document]
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Unnecessary with update-last-pointer support.

File: src/cljam/io/bam/writer.clj [hunk 14]
```diff
@@ -1,23 +1,45 @@
...
+  (:import [java.io Closeable DataOutputStream]
+           [cljam.io.bam.decoder BAMRawBlock]
+           [bgzf4j BGZFOutputStream]))
 
 (declare write-header* write-refs* write-alignments* write-blocks*)
 
+(defn- sort-by-pos?
+  "Returns true if the bam is sorted by coordinate, false if not.
+   It is detected by
+  `@HD SO:coordinate` tag in the header."
```
Comment (alumi) [inline src/cljam/io/bam/writer.clj:21]:
```suggestion
   It is detected by `@HD SO:coordinate` tag in the header."
```

File: src/cljam/io/bam/writer.clj [hunk 15]
```diff
@@ -53,14 +78,46 @@
...
 (defn write-alignments* [^BAMWriter wtr alns header]
-  (let [w (.writer wtr)
+  (let [dw (.data-writer wtr)
+        w ^BGZFOutputStream (.writer wtr)
         refs (refs/make-refs header)]
-    (doseq [a alns]
-      (lsb/write-int w (encoder/get-block-size a))
-      (encoder/encode-alignment w a refs))))
+    (with-open [baos (java.io.ByteArrayOutputStream. 4096)
+                dos (java.io.DataOutputStream. baos)]
```
Comment (alumi) [inline src/cljam/io/bam/writer.clj:85]: Add `ByteArrayOutputStream` to the `import` clause of `ns`

```suggestion
    (with-open [baos (ByteArrayOutputStream. 4096)
                dos (DataOutputStream. baos)]
```

File: src/cljam/io/bam/writer.clj [hunk 16]
```diff
@@ -53,14 +78,46 @@
...
+                               (when @(.index wtr)
+                                 (bam-decoder/decode-pointer-block
+                                  (BAMRawBlock. (.toByteArray baos)
+                                                pointer-beg
+                                                (.getFilePointer w))))))
+                           alns)]
+        (if @(.index wtr)
+          (swap! (.index wtr) bai-writer/merge-index
+                 (bai-writer/make-index* pointer-block))
+          (doall pointer-block))
```
Comment (alumi) [inline src/cljam/io/bam/writer.clj:102]:
```suggestion
          (dorun pointer-block))
```

File: src/cljam/io/bam/writer.clj [hunk 17]
```diff
@@ -53,14 +78,46 @@
...
+                           (when @(.index wtr)
+                             (bam-decoder/decode-pointer-block
+                              (BAMRawBlock. data
+                                            pointer-beg
+                                            (.getFilePointer w))))))
+                       blocks)]
+    (if @(.index wtr)
+      (swap! (.index wtr) bai-writer/merge-index
+             (bai-writer/make-index* pointer-block))
+      (doall pointer-block))
```
Comment (alumi) [inline src/cljam/io/bam/writer.clj:122]:
```suggestion
      (dorun pointer-block))
```

File: src/cljam/io/bam_index/writer.clj [hunk 18]
```diff
@@ -310,7 +228,114 @@
...
+                                  (inc no-coordinate-alns)
+                                  no-coordinate-alns)
+            indices' (if new-ref?
+                       (assoc indices rid idx-status)
+                       indices)]
+        (recur rest rid' idx-status' no-coordinate-alns' indices'))
+      (assoc indices rid idx-status
+             :no-coordinate-alns no-coordinate-alns))))
+
+(defn update-last-pointer [index eof-ptr]
```
Comment (alumi) [inline src/cljam/io/bam_index/writer.clj:290]: Add a docstring

Comment (alumi) [reply src/cljam/io/bam_index/writer.clj:290]: And would you add a unit test for this function?

File: test/cljam/io/sam_test.clj [hunk 19]
```diff
@@ -309,6 +310,43 @@
...
+      (with-open [w (sam/writer temp-bam-file true)]
+        (sam/write-header w (:header test-sam-sorted-by-qname))
+        (sam/write-refs w (:header test-sam-sorted-by-qname))
+        (sam/write-alignments w (:alignments test-sam-sorted-by-qname) (:header test-sam-sorted-by-qname)))
+      (is (not (.exists (cio/file (str temp-bam-file ".bai")))))
+      (with-open [w (sam/writer temp-bam-file true)]
+        (sam/write-header w (:header test-sam-sorted-by-pos))
+        (sam/write-refs w (:header test-sam-sorted-by-pos))
+        (sam/write-alignments w (:alignments test-sam-sorted-by-pos) (:header test-sam-sorted-by-pos)))
+      (is (.exists (cio/file (str temp-bam-file ".bai"))))))
```
Comment (alumi) [inline test/cljam/io/sam_test.clj:344]: Check if the content of the BAI is correct

File: test/cljam/io/sam_test.clj [hunk 20]
```diff
@@ -309,6 +310,43 @@
...
+  (testing "throws Exception"
+    (with-before-after {:before (prepare-cache!)
+                        :after (clean-cache!)}
+      (is (thrown? Exception (sam/writer temp-sam-file true))))))
+
```
Comment (alumi) [inline test/cljam/io/sam_test.clj:349]: Add some test cases for BAI generation with `write-blocks`

File: src/cljam/io/bam_index/writer.clj [hunk 21]
```diff
@@ -253,42 +198,15 @@
...
+;; Making index
+;; -----------
 
 (def ^:dynamic *alignments-partition-size*
   "The number of alignments that is loaded each indexing process. This has an
   effect on performance of concurrent indexing. The default value is 10,000."
   10000)
 
-(defn make-index-from-blocks
+(defn- make-index-from-blocks
```
Comment (alumi) [inline src/cljam/io/bam_index/writer.clj:209]:
```suggestion
(defn make-index-from-blocks
```

File: src/cljam/io/bam/writer.clj [hunk 22]
```diff
@@ -53,14 +78,46 @@
...
+                               (lsb/write-int dw (.size baos))
+                               (.writeTo baos dw)
+                               (when @(.index wtr)
+                                 (bam-decoder/decode-pointer-block
+                                  (BAMRawBlock. (.toByteArray baos)
+                                                pointer-beg
+                                                (.getFilePointer w))))))
+                           alns)]
+        (if @(.index wtr)
+          (swap! (.index wtr) bai-writer/merge-index
```
Comment (alumi) [inline src/cljam/io/bam/writer.clj:100]: I'm so sorry 🤦  We can't use `swap!` for this case as it retains the argument while trying CAS. Would you revert the change, please?

File: src/cljam/io/bam/writer.clj (hunk 14 above)
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj (hunk 15 above)
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj (hunk 16 above)
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj (hunk 17 above)
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: src/cljam/io/bam_index/writer.clj (hunk 21 above)
Comment (xckitahara) [reply src/cljam/io/bam_index/writer.clj]: Fixed.

File: src/cljam/io/bam/writer.clj (hunk 22 above)
Comment (xckitahara) [reply src/cljam/io/bam/writer.clj]: Fixed.

File: test/cljam/io/sam_test.clj (hunk 20 above)
Comment (xckitahara) [reply test/cljam/io/sam_test.clj:359]: Test added.

File: test/cljam/io/sam_test.clj (hunk 19 above)
Comment (xckitahara) [reply test/cljam/io/sam_test.clj]: Test added.

File: src/cljam/io/bam_index/writer.clj (hunk 18 above)
Comment (xckitahara) [reply src/cljam/io/bam_index/writer.clj]: Docstring added.
Test added.

**Comment by xckitahara** [comment]
Performance comparison

Time measurement of sort:
① before change,
② after change (with bai created),
③ after change (without bai created)
It was measured at. (Since sort is used for both aln write and block write)
The target file is medium-sam-file used in the test etc.

① time: 761.079912 ms, sd: 98.673396 µs
② time: 925.187657 ms, sd: 47.099024 µs
③ time: 741.354834 ms, sd: 76.992721 µs
